pkg/controller/projects/projectdeploytokens/controller.go

```diff
@@ -0,0 +1,187 @@
+
+	"github.com/crossplane-contrib/provider-gitlab/apis/projects/v1alpha1"
+	"github.com/crossplane-contrib/provider-gitlab/pkg/clients"
+	"github.com/crossplane-contrib/provider-gitlab/pkg/clients/projects"
+)
+
+const (
+	errNotProjectDeployToken = "managed resource is not a Gitlab deploytoken custom resource"
+	errGetFailed             = "cannot get Gitlab deploytoken"
+	errCreateFailed          = "cannot create Gitlab deploytoken"
+	errDeleteFailed          = "cannot delete Gitlab deploytoken"
+)
+
+// SetupProjectDeployToken adds a controller that reconciles ProjectDeployTokens.
+func SetupProjectDeployToken(mgr ctrl.Manager, l logging.Logger) error {
+	name := managed.ControllerName(v1alpha1.ProjectDeployTokenKind)
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		For(&v1alpha1.ProjectDeployToken{}).
+		Complete(managed.NewReconciler(mgr,
+			resource.ManagedKind(v1alpha1.ProjectDeployTokenGroupVersionKind),
+			managed.WithExternalConnecter(&connector{kube: mgr.GetClient(), newGitlabClientFn: projects.NewProjectDeployTokenClient}),
+			managed.WithInitializers(managed.NewDefaultProviderConfig(mgr.GetClient())),
+			managed.WithLogger(l.WithValues("controller", name)),
+			managed.WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name)))))
+}
+
+type connector struct {
+	kube              client.Client
+	newGitlabClientFn func(cfg clients.Config) projects.ProjectDeployTokenClient
+}
+
+func (c *connector) Connect(ctx context.Context, mg resource.Managed) (managed.ExternalClient, error) {
+	cr, ok := mg.(*v1alpha1.ProjectDeployToken)
+	if !ok {
+		return nil, errors.New(errNotProjectDeployToken)
+	}
+	cfg, err := clients.GetConfig(ctx, c.kube, cr)
+	if err != nil {
+		return nil, err
+	}
+	return &external{kube: c.kube, client: c.newGitlabClientFn(*cfg)}, nil
+}
+
+type external struct {
+	kube   client.Client
+	client projects.ProjectDeployTokenClient
+}
+
+func (e *external) Observe(ctx context.Context, mg resource.Managed) (managed.ExternalObservation, error) {
+	cr, ok := mg.(*v1alpha1.ProjectDeployToken)
+	if !ok {
+		return managed.ExternalObservation{}, errors.New(errNotProjectDeployToken)
+	}
+
+	externalName := meta.GetExternalName(cr)
+	if externalName == "" {
+		return managed.ExternalObservation{ResourceExists: false}, nil
+	}
+
+	projectDeployTokenID, err := strconv.Atoi(externalName)
+	if err != nil {
+		return managed.ExternalObservation{}, errors.New(errNotProjectDeployToken)
+	}
+
+	dt := &gitlab.DeployToken{}
+	listProjectDeployTokensOptions := gitlab.ListProjectDeployTokensOptions{}
+	deploytokenArr, _, err := e.client.ListProjectDeployTokens(*cr.Spec.ForProvider.ProjectID, &listProjectDeployTokensOptions)
+	if err != nil {
+		return managed.ExternalObservation{}, errors.Wrap(resource.Ignore(projects.IsErrorProjectDeployTokenNotFound, err), errGetFailed)
+	}
+
+	for _, v := range deploytokenArr {
+		if v.ID == projectDeployTokenID {
+			dt = v
+			break
```

Comment:
Maybe you could move this code into its own function `findDeployToken` to have cleaner code?

Reply:
yes, I think it makes sense